sdk/storage/AzureStorageBlob/Source/TransferManager.swift

```diff
@@ -82,9 +81,68 @@ public protocol TransferDelegate: AnyObject {
 
 // MARK: Extensions
 
-internal extension TransferManager {
-    func transfer(withId id: UUID) -> Transfer? {
-        return transfers.first { $0.id == id }
+public extension Array where Element == Transfer {
+    /// Retrieve all upload transfers where the source matches the provided source URL.
+    /// - Parameters:
+    ///   - sourceURL: The URL to a file on this device.
+    func from(_ sourceURL: URL) -> [Transfer] {
+        return filter { transfer in
+            guard let transfer = transfer as? BlobTransfer else { return false }
+            return transfer.transferType == .upload && transfer.source == sourceURL
+        }
+    }
+
+    /// Retrieve all download transfers where the source container and blob match the provided parameters.
+    /// - Parameters:
+    ///   - container: The name of the container.
+    ///   - blob: The name of the blob.
+    func from(container: String, blob: String) -> [Transfer] {
+        let pathSuffix = "\(container)/\(blob)"
+        return filter { transfer in
+            guard let transfer = transfer as? BlobTransfer, let source = transfer.source else { return false }
+            return transfer.transferType == .download && source.path.hasSuffix(pathSuffix)
+        }
+    }
+
+    /// Retrieve all download transfers where the destination matches the provided destination URL.
+    /// - Parameters:
+    ///   - destinationURL: The URL to a file path on this device.
+    func to(_ destinationURL: URL) -> [Transfer] {
+        return filter { transfer in
+            guard let transfer = transfer as? BlobTransfer else { return false }
+            return transfer.transferType == .download && transfer.destination == destinationURL
+        }
+    }
+
+    /// Retrieve all upload transfers where the destination container and blob match the provided parameters.
+    /// - Parameters:
+    ///   - container: The name of the container.
+    ///   - blob: The name of the blob.
+    func to(container: String, blob: String) -> [Transfer] {
+        let pathSuffix = "\(container)/\(blob)"
+        return filter { transfer in
+            guard let transfer = transfer as? BlobTransfer, let destination = transfer.destination else { return false }
+            return transfer.transferType == .upload && destination.path.hasSuffix(pathSuffix)
+        }
+    }
+
+    /// Retrieve all transfers of the provided type.
+    ///
+    /// - Parameters:
+    ///   - type: The type of transfers to retrieve.
+    func of(type transferType: TransferType) -> [Transfer] {
```

Comment:
I could be convinced of that. I think the ergonomics of `of(type)` feel a little better but then on the receiving end the customer still has to cast them which does suck a little.

Reply:
To summarize our conversation, `.filter` is a standard Swift idiom for searching collections, so if we can standardize these methods on the `.filter(...)` paradigm we should be good to go. As for what convenience methods we expose, I'd say make your best judgement. We can always change them based on customer feedback instead of my internal philosophy :-P

Reply:
I took a different approach, which removed the need for the `of(type:)` method - let me know what you think. I feel like the API is much more understandable now from the consumer's POV. But if you feel like there's still a better approach, ping me and we can discuss further - don't feel like you have to sign off just to get this merged.